File: core/caption_compliance.py
"""Deterministic caption compliance validator shared by save/approval/preflight logic."""
from __future__ import annotations

import re
from typing import Any


def _text(value: Any) -> str:
    return str(value or "")


def _contains(text: str, value: str) -> bool:
    return _text(value).casefold() in text.casefold()


def validate_caption_revision(revision: dict[str, Any], profile: dict[str, Any], *, expected_rules_hash: str | None = None) -> dict[str, Any]:
    text = _text(revision.get("text"))
    failures: list[dict[str, Any]] = []
    def fail(field: str, code: str, expected: Any, actual: Any = None) -> None:
        failures.append({"field": field, "code": code, "expected": expected, "actual": actual})

    for handle in profile.get("required_handles") or []:
        if not _contains(text, _text(handle)):
            fail("handles", "required_handle_missing", handle)
    for hashtag in profile.get("required_hashtags") or []:
        if not _contains(text, _text(hashtag)):
            fail("hashtags", "required_hashtag_missing", hashtag)
    for disclosure in profile.get("required_disclosures") or []:
        if not _contains(text, _text(disclosure)):
            fail("disclosures", "required_disclosure_missing", disclosure)
    for phrase in profile.get("required_phrases") or []:
        if not _contains(text, _text(phrase)):
            fail("phrases", "required_phrase_missing", phrase)
    for cta in profile.get("cta") or []:
        if not _contains(text, _text(cta)):
            fail("cta", "cta_missing", cta)
    for term in profile.get("prohibited_terms") or []:
        if _contains(text, _text(term)):
            fail("prohibited_terms", "prohibited_term_present", term)
    limit = int(profile.get("caption_limit") or 0)
    if limit and len(text) > limit:
        fail("text", "caption_too_long", limit, len(text))
    if expected_rules_hash is not None and revision.get("rules_hash") != expected_rules_hash:
        fail("rules_hash", "rules_hash_mismatch", expected_rules_hash, revision.get("rules_hash"))
    if revision.get("caption_hash"):
        from core.caption_revisions import revision_hash
        comparable = dict(revision)
        comparable.pop("caption_hash", None)
        comparable.pop("revision_id", None)
        if revision_hash(comparable) != revision.get("caption_hash"):
            fail("caption_hash", "caption_hash_mismatch", revision.get("caption_hash"))
    return {"ok": not failures, "status": "pass" if not failures else "fail", "platform": profile.get("platform"), "profile_version": profile.get("version"), "failures": failures}


__all__ = ["validate_caption_revision"]

File: cloudflare/caption_compliance.js
export function validateCaptionRevision(revision, profile, expectedRulesHash = null) {
  const text = String(revision?.text || "");
  const failures = [];
  const contains = (value) => text.toLocaleLowerCase().includes(String(value || "").toLocaleLowerCase());
  const fail = (field, code, expected, actual = null) => failures.push({ field, code, expected, actual });
  for (const value of profile?.required_handles || []) if (!contains(value)) fail("handles", "required_handle_missing", value);
  for (const value of profile?.required_hashtags || []) if (!contains(value)) fail("hashtags", "required_hashtag_missing", value);
  for (const value of profile?.required_disclosures || []) if (!contains(value)) fail("disclosures", "required_disclosure_missing", value);
  for (const value of profile?.required_phrases || []) if (!contains(value)) fail("phrases", "required_phrase_missing", value);
  for (const value of profile?.cta || []) if (!contains(value)) fail("cta", "cta_missing", value);
  for (const value of profile?.prohibited_terms || []) if (contains(value)) fail("prohibited_terms", "prohibited_term_present", value);
  const limit = Number(profile?.caption_limit || 0);
  if (limit && text.length > limit) fail("text", "caption_too_long", limit, text.length);
  if (expectedRulesHash !== null && revision?.rules_hash !== expectedRulesHash) fail("rules_hash", "rules_hash_mismatch", expectedRulesHash, revision?.rules_hash);
  return { ok: failures.length === 0, status: failures.length ? "fail" : "pass", platform: profile?.platform, profile_version: profile?.version, failures };
}
